plutus-benchmark/script-contexts/src/PlutusBenchmark/Data/ScriptContexts.hs

```diff
@@ -0,0 +1,148 @@
+import PlutusLedgerApi.V1.Data.Value
+import PlutusTx qualified
+import PlutusTx.Builtins qualified as PlutusTx
+import PlutusTx.Data.AssocMap qualified as Map
+import PlutusTx.Plugin ()
+import PlutusTx.Prelude qualified as PlutusTx
+
+-- | A very crude deterministic generator for 'ScriptContext's with size
+-- approximately proportional to the input integer.
+mkScriptContext :: Int -> ScriptContext
+mkScriptContext i =
+  ScriptContext
+    (mkTxInfo i)
+    (Redeemer (PlutusTx.toBuiltinData (1 :: Integer)))
+    (SpendingScript (TxOutRef (TxId "") 0) Nothing)
+
+
+mkTxInfo :: Int -> TxInfo
+mkTxInfo i = TxInfo {
+  txInfoInputs=mempty,
+  txInfoReferenceInputs=mempty,
+  txInfoOutputs=fmap mkTxOut [1..i],
+  txInfoFee=10000,
+  txInfoMint=mempty,
+  txInfoTxCerts=mempty,
+  txInfoWdrl=Map.empty,
+  txInfoValidRange=always,
+  txInfoSignatories=mempty,
+  txInfoRedeemers=Map.empty,
+  txInfoData=Map.empty,
+  txInfoId=TxId "",
+  txInfoVotes=Map.empty,
+  txInfoProposalProcedures=mempty,
+  txInfoCurrentTreasuryAmount=Nothing,
+  txInfoTreasuryDonation=Nothing
+  }
+
+mkTxOut :: Int -> TxOut
+mkTxOut i = TxOut {
+  txOutAddress=pubKeyHashAddress (PubKeyHash ""),
+  txOutValue=mkValue i,
+  txOutDatum=NoOutputDatum,
+  txOutReferenceScript=Nothing
+  }
+
+mkValue :: Int -> Value
+mkValue i = assetClassValue (assetClass adaSymbol adaToken) (fromIntegral i)
+
+-- This example decodes the script context (which is O(size-of-context) work), and then also
+-- does some work that's roughly proportional to the size of the script context (counting the
+-- outputs). This should be a somewhat realistic example where a reasonable chunk of work is
+-- done in addition to decoding.
+{-# INLINABLE checkScriptContext1 #-}
+checkScriptContext1 :: PlutusTx.BuiltinData -> ()
+checkScriptContext1 d =
+  -- Bang pattern to ensure this is forced, probably not necesssary
+  -- since we do use it later
+  let !sc = PlutusTx.unsafeFromBuiltinData d
+      ScriptContext txi _ _ = sc
+  in
+  if PlutusTx.length (txInfoOutputs txi) `PlutusTx.modInteger` 2 PlutusTx.== 0
+  then ()
+  else PlutusTx.traceError "Odd number of outputs"
+
+mkCheckScriptContext1Code :: ScriptContext -> PlutusTx.CompiledCode ()
+mkCheckScriptContext1Code sc =
+  let d = PlutusTx.toBuiltinData sc
+  in
+    $$(PlutusTx.compile [|| checkScriptContext1 ||])
+    `PlutusTx.unsafeApplyCode`
+    PlutusTx.liftCodeDef d
+
+-- This example aims to *force* the decoding of the script context and then ignore it entirely.
+-- This corresponds to the unfortunate case where the decoding "wrapper" around a script forces
+-- all the decoding work to be done even if it isn't used.
+{-# INLINABLE checkScriptContext2 #-}
+checkScriptContext2 :: PlutusTx.BuiltinData -> ()
+checkScriptContext2 d =
+  let (sc :: ScriptContext) = PlutusTx.unsafeFromBuiltinData d
+  -- Just using a bang pattern was not enough to stop GHC from getting
+  -- rid of the dead binding before we even hit the plugin, this works
+  -- for now!
```

Comment:
Did you consider deriving `NFData` instances and using `deepseq` to deeply evaluate the script context?

Reply:
This is basically all copied over from `PlutusBenchmark.ScriptContexts`, and most of the documentation in the other modules is as well copied over from the non-data versions. Which reminds me that I need properly document things, but once I refactor and remove the duplication.